### PR TITLE
Analyzer: Revise - Abstract RuleBase from DataDiagnosis

### PR DESCRIPTION
**Pull request by yukirora**
**Description**
Abstract RuleBase from DataDiagnosis.

### CODE REVIEW

File: superbench/analyzer/rule_base.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
+# Copyright (c) Microsoft Corporation.
```
Comment (cp5555) [inline superbench/analyzer/rule_base.py:1]: Do we need to add test for `rule_base.py`?

**Review by cp5555 — APPROVED** [review]

File: superbench/analyzer/rule_base.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+class RuleBase():
+    """RuleBase class."""
+    def __init__(self):
+        """Init function."""
+        self._sb_rules = {}
+        self._benchmark_metrics_dict = {}
+
+    def _get_metrics_by_benchmarks(self, metrics_list):
+        """Get mappings of benchmarks:metrics of metrics_list.
```
Comment (guoshzhao) [inline superbench/analyzer/rule_base.py:20]: `Get mappings of benchmarks:metrics of metrics_list.` => `Get mappings of benchmarks:metrics from metrics_list.`

File: superbench/analyzer/rule_base.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+                    self._sb_rules[rule]['metrics'][metric] = -1
+                    self._enable_metrics.add(metric)
+                    continue
+                # metric full name not in baseline, use regex to match
+                for metric_regex in benchmark_metrics_dict_in_rule[benchmark_name]:
+                    if re.search(metric_regex, metric):
+                        self._sb_rules[rule]['metrics'][metric] = -1
+                        self._enable_metrics.add(metric)
+
+    def preprocess(self, raw_data_file, rule_file):
```
Comment (guoshzhao) [inline superbench/analyzer/rule_base.py:86]: public function - without underline
protect function - one underline
private function - two underline

Could you please rename these functions?

File: superbench/analyzer/rule_base.py (hunk 2 above)
Comment (yukirora) [reply superbench/analyzer/rule_base.py]: thanks, updated

File: superbench/analyzer/rule_base.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+                logger.warning('RuleBase: get_metrics_by_benchmarks - {} does not have benchmark_name'.format(metric))
+            else:
+                benchmark = metric.split('/')[0]
+                if benchmark not in benchmarks_metrics:
+                    benchmarks_metrics[benchmark] = set()
+                benchmarks_metrics[benchmark].add(metric)
+        return benchmarks_metrics
+
+    def _check_rules(self, rule, name):
+        """Check the rule of the metric whether the formart is valid.
```
[the remark this answers is not in the document]
Comment (yukirora) [reply superbench/analyzer/rule_base.py]: yes, changed to _check_rules_and_format?

File: superbench/analyzer/rule_base.py (hunk 3 above)
Comment (yukirora) [reply superbench/analyzer/rule_base.py]: updated to _preprocess

**Review by guoshzhao — APPROVED** [review]

File: superbench/analyzer/rule_base.py (hunk 1 above)
Comment (yukirora) [reply superbench/analyzer/rule_base.py:1]: i think the functions in rule_base have been tested in the test for DataDiagnosis

**Pull request merged**